doc/storage/api/api.rst

```diff
@@ -658,15 +842,193 @@ These definitions must be defined in the header file :file:`psa/protected_storag
+
+   A caller may call ``psa_ps_rename(uid, uid_new)`` without invalidating the iteration context. The iterator must not return ``uid``. The iterator is not guaranteed to return ``uid_new``, the behavior is dependent on both implementation and identity.
+
+   The following code snippet uses a linked list to store the matching files before iterating over that list and removing them.
+
+   .. code-block:: c
+
+      my_context = NULL;
+      my_filter = 0x1111 0000 0000 0000;
+      my_length = 0x0020;
+      my_result = NULL;
+      if psa_ps_iterator_start(my_context, my_filter, my-length, my_result) == PSA_SUCCESS
+      	{
+      	do
+      	   {
+      	   	// do something with my_result
+      	    psa_ps_iterator_next(my_context, my_result)
+      	    // we will get an 'does not exist error' when we reach the last item, any other error is a storage failure
+      	    if my_reult <> PSA_ERROR_DOES_NOT_EXIST
+      	   	  {
+      	   	  	/* deal with storage failure */
+      	   	  }
+      	   }
+        while my_result == PSA_SUCCESS ;
+        };
+
+
+
+
+.. function:: psa_ps_iterator_next
+
+   .. summary::
+      Returns the next ``uid`` in this iteration.
+      This function must be fully defined if iteration is supported.
+      If iteration is not supported, then this function SHALL be defined in the header - potentially as an inline function - but SHALL return ``PSA_ERROR_NOT_SUPPORTED``
+
+   .. param::  psa_ps_storage_iterator_t* context
+       A pointer to a context for this iterator as returned by `psa_ps_iterator_start` or updated by a previous call to `psa_ps_iterator_next`.  The content of the iterator will change on success and is undefined on error.
+
+   .. param:: psa_storage_uid_t *result
+        A pointer to the location in which to store ``uid``. On success the contents of this location will be updated with the next matching ``uid``. On error, the contents are undefined.
+
+   .. return:: psa_status_t
+       A status indicating the success or failure of the operation.
+
+   .. retval:: PSA_SUCCESS
+       The operation completed successfully.
+
+   .. retval:: PSA_ERROR_DOES_NOT_EXIST
+       The iterator has returned all the uids that match this iteration.
+
+   .. retval:: PSA_ERROR_STORAGE_FAILURE
+       The operation failed because the physical storage has failed (Fatal error).
+
+   .. retval:: PSA_ERROR_DATA_CORRUPT
+       The operation failed because the contents of the iteration have changed. That is a ``uid`` matching the filter has either been created or deleted.
+
+   .. retval:: PSA_ERROR_INVALID_ARGUMENT
+       The operation failed because either:
+
+       * The provided context is not valid.
+
+       * The caller cannot access the memory at ``result``
+
```

Comment:
Probably worth at least a single line in the function description to refer to the start function description of how the iterator works and how to use it.